### PR TITLE
[Android] Fixed local images in production builds (Fixes #273, #296 & #246)

### LINKED ISSUE #273: Can not show local Image
I have a problem When I user <Image />
 Somes times I load an local Image eg: <Image source={require('../images/Icon_thunder.png')}/> ,So I had Substitute <FastImage /> for <Image />
But Recently，Our online version still has many problems that cannot be loaded with local images.

### PR DESCRIPTION
**Pull request by patrickkempff**
This PR fixes the passing of require in production builds.

Context:
When using the following component: `<FastImage source={require('./image.png')} />`. The image loads just fine when running the android debug build. But when running the production variant, the same component is broken. This is because in debug mode the result of resolveAssetSource
is a `http://`resource as it want to fetch the image from the packager. In production a local drawable name is returned so it can be loaded from the Drawable folder in the APK itself. At the moment FastImage does not handle this case correctly.

Fix:
Refactored the mechanism that resolves the resource uri that Glide can understand. I investigated how the stock RN Image component handles this case, as it turns out `ImageSource` is used to transform the output of resolveAssetSource to a local `res:/` uri. We then take this uri and transform to a format that Glide understands (android.resource://). This will be handle by `FastImageSource` which is build on top of `ImageSource`

This PR is build on top of #272

Done:
- [x] Support require() in production mode (Fixes #273, #296 & #246)
- [x] Fix crash when passing null uri
- [x] Prevents crash when activity is already destroyed
- [x] Fixes possible breakage with newer RN version by using `Image.resolveAssetSource` instead of importing it directly as it is not offically supported.
- [x] Make sure `data:image/png;base64` uri's still work.
- [x] Added missing breaks in switch statement for caching strategy, thanks @Yria! (#282)

### CODE REVIEW

**Comment by manhhoan73** [comment]
Can you give me the example? I build the release not display the image.

**Comment by fierysolid** [comment]
@patrickkempff does this add support for bundled images? Documentation on bundled images can be found [here](https://facebook.github.io/react-native/docs/images#images-from-hybrid-app-s-resources)

**Comment by patrickkempff** [comment]
@fierysolid not at the moment, I am working on a solution to support this.

> @patrickkempff does this add support for bundled images? Documentation on bundled images can be found [here](https://facebook.github.io/react-native/docs/images#images-from-hybrid-app-s-resources)

**Comment by patrickkempff** [comment]
@manhhoan73 can you tell me the steps you took integrating this PR?

**Comment by DylanVann** [comment]
screen shot 2018-10-22 at 10 59 15 pm

😢

File: src/index.js [hunk 1]
```diff
@@ -34,7 +32,7 @@ class FastImage extends Component {
             ...props
         } = this.props
 
-        const resolvedSource = resolveAssetSource(source)
+        const resolvedSource = Image.resolveAssetSource(source)
```
Comment (DylanVann) [inline src/index.js]: Awesome. It was not a nice looking import before.

File: android/src/main/java/com/dylanvann/fastimage/FastImageViewManager.java [hunk 2]
```diff
@@ -153,4 +182,50 @@ public float getGranularityPercentage() {
...
+    private static boolean isAcitityDestroyed (Activity activity) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+            return activity.isDestroyed() || activity.isFinishing();
+        } else {
+            return activity.isFinishing() || activity.isChangingConfigurations();
+        }
+
+    }
+
+    private void warnImageSource(Context context, String uri) {
```
Comment (DylanVann) [inline android/src/main/java/com/dylanvann/fastimage/FastImageViewManager.java]: Doesn't seem to be used.

**Comment by DylanVann** [comment]
I resolved some of the conflicts and merged this in another PR (https://github.com/DylanVann/react-native-fast-image/pull/324).

Need to do some more testing before another release, but it looks good. Thank you!

**Pull request closed without merging**